File: algorithm/baekjoon/src/main/java/com/keunwon/algorithm/bruteforce/Problem15663.java
package com.keunwon.algorithm.bruteforce;

import java.io.BufferedWriter;
import java.io.OutputStreamWriter;
import java.util.*;

public class Problem15663 {
    public static final BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

    public static int[] numbers;
    public static int[] result;
    public static boolean[] visited;
    public static Set<String> set = new LinkedHashSet<>();

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        final int N = sc.nextInt();
        final int M = sc.nextInt();

        numbers = new int[N];
        visited = new boolean[N];
        result = new int[M];

        for (int i = 0; i < N; i++) {
            numbers[i] = sc.nextInt();
        }

        Arrays.sort(numbers);
        dfs(N, M, 0);
        set.forEach(System.out::println);
    }

    static void dfs(int n, int m, int depth) {
        if (m == depth) {
            StringBuilder sb = new StringBuilder();
            for (int i : result) {
                sb.append(i).append(" ");
            }
            set.add(sb.toString());
            return;
        }

        for (int i = 0; i < n; i++) {
            if (visited[i]) { continue; }

            visited[i] = true;
            result[depth] = numbers[i];
            dfs(n, m, depth + 1);
            visited[i] = false;
        }
    }
}

File: algorithm/kotlin-algorithm/src/main/kotlin/com/keunwon/algorithm/baekjoon/Problem15663.kt
package com.keunwon.algorithm.baekjoon

/**
 * Title: N과 M (9)
 * Level: 실버-2
 **/
class Problem15663 {
    private val answer = mutableListOf<String>()

    fun solution(m: Int, numbers: IntArray): List<String> {
        numbers.sort()

        dfs(0, BooleanArray(numbers.size), IntArray(m), numbers)
        return answer.distinct()
    }

    private fun dfs(
        depth: Int,
        visited: BooleanArray,
        picks: IntArray,
        numbers: IntArray,
    ) {
        if (depth == picks.size) {
            answer.add(picks.joinToString(" "))
            return
        }

        for ((i, num) in numbers.withIndex()) {
            if (visited[i]) continue

            visited[i] = true
            picks[depth] = num
            dfs(depth + 1, visited, picks, numbers)
            visited[i] = false
        }
    }
}

fun main() = with(System.`in`.bufferedReader()) {
    val bw = System.out.bufferedWriter()

    val (n, m) = readLine().split(" ").map { it.toInt() }
    val numbers = readLine().split(" ").map { it.toInt() }.toIntArray()

    Problem15663().solution(m, numbers).forEach {
        bw.write(it)
        bw.newLine()
    }

    bw.flush()
    bw.close()
    close()
}
